### ISSUE TITLE
v27.0.0 breaks TypeScript typings

### ISSUE THREAD

**Issue by hehex9** [primer/react#1163]
**Describe the bug**
1. some components (Button*/Dialog/TextInput/FilterList/...) have incorrect type definitions, and ts keeps complaining:
```sh
Property 'css' is missing in type '{}' but required in type 'Pick<TextInputInternalProps, "form" | "p" | "slot" | "style" | "title" | "pattern" | "children" | "type" | "theme" | "key" | "defaultChecked" | "defaultValue" | ... 317 more ... | "icon">'.ts(2741)
```
2. callbacks like `onChange`/`onKeyPress` in `TextInput` are not giving us the right types, it will cause an error like:
```sh
Parameter 'e' implicitly has an 'any' type.ts(7006)
```

vscode shows the `onChange`'s type definition is:
```ts
onChange?: (((event: React.FormEvent<HTMLSpanElement>) => void) & ((event: React.ChangeEvent<HTMLInputElement>) => void)) | undefined
```
which is not accurate

**To Reproduce**
[codesanbox](https://codesandbox.io/s/primer-components-ts-mcsfk?file=/src/index.tsx)

**Expected behavior**
no error

**Screenshots**
<img width='420' alt='codesanbox' src='https://user-images.githubusercontent.com/9209882/114643583-8618bf00-9d08-11eb-9029-c54b89fff6fc.jpg' />

**Additional context**
v26 was fine

**Comment by colebemis** [comment]
Thanks for the bug report and clear reproduction, @hehex9! We'll investigate this week.

**Comment by hehex9** [comment]
> we're not seeing these TypeScript errors in any of the internal projects

That's weird, since it's pretty easy to reproduce in codesandbox i think :(

I'll look deeper into it when I have time, would you mind sharing your attempts here? Thanks

**Comment by velocity23** [comment]
Having issues as well, having to `@ts-ignore` for `TextInput`

**Comment by yareyaredesuyo** [comment]
Maybe related to using React.ComponentType?
I think there is no need to use it, just declare type and export it.

Refactor like this (using Box):

before

```tsx
const Box = styled.div<SystemCommonProps & SystemFlexProps & SystemLayoutProps & SxProp>`
  ${COMMON}
  ${FLEX}
  ${LAYOUT}
  ${sx};
`

export type BoxProps = ComponentProps<typeof Box>
```

after

```tsx
export type BoxProps = SystemCommonProps & SystemFlexProps & SystemLayoutProps & SxProp

const Box = styled.div<BoxProps>`
  ${COMMON}
  ${FLEX}
  ${LAYOUT}
  ${sx};
`
```

**Comment by ludwigschubi** [comment]
Having the same error after bootstrapping a cra with template typescript and adding @primer-components

App.tsx looks like:
```typescript
import React from "react";
import { ThemeProvider, Button } from "@primer/components";

// import styles from "./App.module.css";

export const App: React.FC = () => {
  return (
    <ThemeProvider>
      <p>
        Edit <code>src/App.tsx</code> and save to reload.
      </p>
      <Button>
        <a href="https://reactjs.org" target="_blank" rel="noopener noreferrer">
          Learn React
        </a>
      </Button>
    </ThemeProvider>
  );
};

export default App;
```

Error only disappears after adding css prop to <Button />

**Comment by MichalKowalczyk** [comment]
@dmarcey 
Your solution looks nice, but when I'm trying implement it, I'm getting TypeScript error on every JSX elements:

`JSX element class does not support attributes because it does not have a 'props' property.`

**Comment by jfuchs** [comment]
Hey @Mikescops, it's on our "up next" list, but work hasn't begun and I don't have a time estimate. Were you able to try the workaround described above?

**Comment by Mikescops** [comment]
I had much more typing issues with the workaround, so it wasn't really helping. I'll look forward to your fix, no problem, thanks!

**Comment by gregarnott** [comment]
I recognise this issue from `styled-components` and `@types/styled-components` and similar issues.
Taken from https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31245

To prevent TypeScript errors on the css prop on arbitrary elements, add the following:

```ts
    import {} from 'styled-components/cssprop'
```

This only needs to be done once in the project.

**Comment by jfuchs** [comment]
The fix is now published in 29.1.1. Thank you for your patience.
